src/iceberg/expression/manifest_evaluator.cc

```diff
@@ -0,0 +1,390 @@
+      return Invalid("Invalid literal: not a string, cannot use StartsWith");
+    }
+    const auto& prefix = std::get<std::string>(lit.value());
+    ICEBERG_ASSIGN_OR_RAISE(
+        auto lower, DeserializeBoundLiteral(summary.lower_bound.value(), ref->type()));
+    ICEBERG_ASSIGN_OR_RAISE(
+        auto upper, DeserializeBoundLiteral(summary.upper_bound.value(), ref->type()));
+    const auto& lower_bound = std::get<std::string>(lower.value());
+    const auto& upper_bound = std::get<std::string>(upper.value());
+    // truncate lower bound so that its length in bytes is not greater than the length of
+    // prefix
+    size_t length = std::min(prefix.size(), lower_bound.size());
+    if (lower_bound.substr(0, length) > prefix) {
+      return kRowCannotMatch;
+    }
+    length = std::min(prefix.size(), upper_bound.size());
+    if (upper_bound.substr(0, length) < prefix) {
+      return kRowCannotMatch;
+    }
+    return kRowsMightMatch;
+  }
+
+  Result<bool> NotStartsWith(const std::shared_ptr<Bound>& expr,
+                             const Literal& lit) override {
+    ICEBERG_ASSIGN_OR_RAISE(auto ref, ParseBoundReference(expr));
+    ICEBERG_ASSIGN_OR_RAISE(auto pos, GetPosition(*ref));
+    const auto& summary = stats_.at(pos);
+    if (summary.contains_null || !summary.lower_bound.has_value() ||
+        !summary.upper_bound.has_value()) {
+      return kRowsMightMatch;
+    }
+    if (lit.type()->type_id() != TypeId::kString) {
+      return Invalid("Invalid literal: not a string, cannot use notStartsWith");
+    }
+    // notStartsWith will match unless all values must start with the prefix. This happens
+    // when the lower and upper bounds both start with the prefix.
+    const auto& prefix = std::get<std::string>(lit.value());
+    ICEBERG_ASSIGN_OR_RAISE(
+        auto lower, DeserializeBoundLiteral(summary.lower_bound.value(), ref->type()));
+    ICEBERG_ASSIGN_OR_RAISE(
+        auto upper, DeserializeBoundLiteral(summary.upper_bound.value(), ref->type()));
+    const auto& lower_bound = std::get<std::string>(lower.value());
+    const auto& upper_bound = std::get<std::string>(upper.value());
+
+    // if lower is shorter than the prefix, it can't start with the prefix
+    if (lower_bound.size() < prefix.size()) {
+      return kRowsMightMatch;
+    }
+    if (lower_bound.starts_with(prefix)) {
+      // the lower bound starts with the prefix; check the upper bound
+      // if upper is shorter than the prefix, it can't start with the prefix
+      if (upper_bound.size() < prefix.size()) {
+        return kRowsMightMatch;
+      }
+      // truncate upper bound so that its length in bytes is not greater than the length
+      // of prefix
+      if (upper_bound.starts_with(prefix)) {
+        return kRowCannotMatch;
+      }
+    }
+    return kRowsMightMatch;
+  }
+
+ private:
+  Result<BoundReference*> ParseBoundReference(const std::shared_ptr<Bound>& expr) const {
```

Comment:
We don't need this. The Java impl extends `BoundExpressionVisitor` which can only deal with bound reference. However, we extend `BoundVisitor` so that we can directly call `expr->reference()` to get a bound reference.